### ISSUE TITLE
new: modal

### ISSUE THREAD

**Issue by jackreeves** [telus/tds-community#61]
## Problem Statement

As a designer i would like a consistent way to show modals so that i have don't have to reinvent the wheel

As a customer i would like a consistent modal interface so i know where to close, click next etc.

## Out of scope

* B2B style dialogue chooser ([see design](https://telus.invisionapp.com/d/main#/console/17778239/387938037/preview))

## Outcome
- community-modal has two interface styles: modal, and dialogue
  - modal should have a confirm and close[x] button
  - dialogue should have a confirm and cancel button
- Consumers of community-modal should handle focus state and return to the initiator of the modal once the modal is cancelled/closed
  - Modal heading should be announced once it receives `focus()`
- Appears over-top parent page with an overlay
- Interactions (skip links etc.) need to remain inside window until closed, with Esc key available to user (accessibility). Esc key press should close the modal.
- Expectations should be set before showing modal (shouldn’t be a surprise)
  - Modal should be announced to assistive tech before it is triggered

**Usage:**
- Triggered based on user action
- Needs immediate action by user
    - Confirmation of action that isn’t reversible
    - Transactionwith an exchange of monetary value, etc.
- Delivering information that requires focus

**Considerations:**
- Does the interaction need to be disruptive
- Can the information be included on the same or subsequent screen?
- Can a required action be include as part of the current flow

#### Summary

A modal is used to reveal additional information to a user after they have performed an explicit interaction. They are a strongly discouraged pattern; it's preferred to have all relevant information within a page, and irrelevant information either linked externally or omitted. 

#### Usage criteria

- Modals must only appear after a user interaction, not on page load or any other circumstance
- Modals must only be triggered by a button 
- Modals must only contain copy and no other content types
- Should keep copy to a minimum
- Refer to [WebAIM documentation](https://www.w3.org/TR/wai-aria-practices/examples/dialog-modal/dialog.html) for how to achieve modals accessibly

#### Accessibility

- When displayed, the modal focus traps the user, and locks page scroll.
- It is scrollable when the viewport height is shorter than the contents of the modal

### Coded component

**Proposed API**
```jsx
<Modal
  heading="Business marketing"
  body="Solutions for your business, TELUS helps market your products."
  cta={<>Sign up <A11yContent>for business marketing</A11yContent></>}
  cancel="No thanks"
/>

heading: string, required
body: string, required
cta: node, required
cancel: node, optional
```

Scenario: modal is not open

- a button can be given a callback function that triggers the modal view
- when the modal initiator is triggered, it passes its `ref` to the Modal component in order to return focus

Scenario: modal is open (updated March 24, 2020)

- For content models; Modal should be able to be closed with a close button (available with `@tds/core-interactive-icon`) on the top-right corner, clicking outside the modal, hitting the ESCAPE key
- For dialogue models, Modal should be able to be cancelled with a dedicated cancel button
  - Modal vs dialogue: the modal component can either render a CTA + close button (modal), or a CTA + cancel button (dialogue)
- When the modal is open, the first text item should be focused and read, as per accessibility standards (set `tabindex="-1"` to the heading)
- modal has header text, body text, required 'submit' text, and optional 'cancel' text
- Modal is focus trapped (use `@tds/util-shared-hocs` `withFocusTrap`)

Scenario: modal styles

- header text renders as an h3 (using a `p` tag) (dependency, need to render core-heading as a `<p>`, can use `div` for the interim)
  - body text renders as a Paragraph, 16 pixels below the header
  - submit text renders as a primary Button, with Box-5 spacing below body text
  - cancel text renders as an outlined Button (currently not in TDS)
- There should be an Athens Grey `#F7F7F8` overlay at 94% opacity, covering the entire page behind the modal
- the modal should have a `box-shadow: 0 2px 4px 0 rgba(0,0,0,0.5);`, which is different than Card
- the modal should have a 1px shuttle grey border (have it match the styles of `@tds/core-card`'s `defaultWithBorder` variant)
- its z-index is `2000` (assuming chatbot has a z-index of `1001`)
- the last button has Box-4 space between the bottom of the modal and button

Situation: responsiveness

- On mobile viewports widths (from xs to sm) (below 576px)
  - the modal must cover the entire viewport, and must not have a box-shadow or border
  - the buttons should be positioned at the bottom (using a flexbox)
  - on short heights (where modal content is larger than the viewport height), the modal should be able to scroll
- On larger viewports (sm+)
  - Max width: it should not exceed 570px in width
  - on short heights, the modal should be able to scroll, without the <body> scrolling, and the modal should only scroll up until the top and bottom-most parts (aka, do not scroll with the entire page height, only the viewport height). You can take inspiration from [Spinner](https://github.com/telus/tds-core/blob/e9a07e7144bcdd90eaa8bc6424146aab05bc301c/packages/Spinner/Spinner.jsx#L53-L61)

#### Examples

Example 1: triggering and displaying a modal

## Designs

https://telus.invisionapp.com/project/17778239

## Meta

* Willing to design solution: No
* Willing to develop solution: No
* Has workaround: No
* Do any similar components already exist: No
* Does this pattern exist already but isn’t a community component? (include url) 

* Which teams/applications do you think would use this improvement?
Myaccount HS MOB B2B
* High impact: No

**Comment by DanGenuardi** [comment]
Has anyone designed an element for this overlay modal? I'm asking because I'm exploring the possibility to use it within the My TELUS experience for the usage page for when someone has to set a limit for a specific subscriber within a shared plan.

**Comment by varunj90** [comment]
hey @DanGenuardi assigning this ticket you based on our conversation on slack! thanks for taking this up and keeping @philroberto research into consideration.

**Comment by DanGenuardi** [comment]
@philroberto 

Here is a basic overlay modal that we coded here for team Nutella. Can you have a look at it and see if there is anything that you may want polished?

<img width="1440" alt="screen shot 2018-12-18 at 1 32 36 pm" src="https://user-images.githubusercontent.com/39997967/50176332-1bd4e880-02cd-11e9-9eaf-be92bab8d73e.png">

**Comment by philroberto** [comment]
@gkychow could you take a look at the design above for modals and provide feedback if any.

@DanGenuardi just a couple things i noticed:

1 - In this context, be sure to reference the requirement "Every modal should have a cancel/exit option", as cancel could provide a clearly stated way to reverse this decision (relying on "close" alone, sets unclear expectations)

2 - Is the white-out overlay standard for MyTelus? I don't anticipate that the outline will be present on the modal, so a dark overlay might make more sense and create a clearer point of focus for the user. @gkychow perhaps we have a dark gradient (alpha / colour) in mind?

Note: Perhaps it's out of our control, but the "Feedback" and "Chat" are still in focus, must that occur? (it suggests to the user that it's part of what should be anticipated regarding selection / focus state.

**Comment by DanGenuardi** [comment]
@philroberto This is just a quick coded design that the Nutella team did a few months back. Just wanted to see what you thought of the white background and the card treatment. I'm still working on the overall design with the elements that you are looking for. You will see something new in the new year.

**Comment by DanGenuardi** [comment]
![screen shot 2019-01-10 at 2 50 12 pm](https://user-images.githubusercontent.com/39997967/50993210-142cec00-14e7-11e9-9311-3aa1bee062be.png)

@philroberto @varunj90 
I have created a mockup on how the overlay modal can look like for TDS.

I made the background colour shark grey so we wouldn’t have to add any new design treatment to the modal box itself. It’s a standard white box so it can stand out more.

I put 32 pixels of padding around the edges for the content inside of the box so it wouldn’t look so crammed in. That would convert to 24 pixels of padding for mobile layouts.

I added a "X" icon which work as a close button at the top of the right hand corner so customers know how to exit the modal experience.

I laid out the content left justified for desktop and tablet view so it’s a cleaner look for customers to read through the content and text fields for them to fill out.

Please let me know what you think, and what are the next steps.

**Comment by DanGenuardi** [comment]
![screen shot 2019-01-15 at 12 31 41 pm](https://user-images.githubusercontent.com/39997967/51198186-8a539900-18c1-11e9-8b74-1ad87638fced.png)

Here is mobile layout for the modal.

**Comment by DanGenuardi** [comment]
![screen shot 2019-01-15 at 1 40 06 pm](https://user-images.githubusercontent.com/39997967/51201855-2cc44a00-18cb-11e9-8ede-bf1275d26880.png)
![screen shot 2019-01-15 at 1 40 16 pm](https://user-images.githubusercontent.com/39997967/51201856-2cc44a00-18cb-11e9-965c-09fb382a849b.png)
![screen shot 2019-01-15 at 1 40 24 pm](https://user-images.githubusercontent.com/39997967/51201863-32219480-18cb-11e9-85d3-95e0198709ac.png)

Made the "X" shark grey.

**Comment by DJDA** [comment]
Hi all - we are also exploring use of a modal for Login for the new renewal flow.  See this clip for a demo of the interation:

**Comment by twirlingsky** [comment]
@DanGenuardi @DJDA 

I have a couple of suggestions for a design revision over what is set in the problem statement:

- add close to the left of the x icon and link the word close. Both need to be purple to demonstrate that they are links
- try rounded corners on the modal (4px)
- bkg overlay, use shark grey at 60% opacity
- add guides to show box spacing
- ensure accessibility (before the modal is prompted the dialog is announce by assistive technology)

Is it possible to make the controls more tied to what is happening? For example, could the CTA say Save data limit, and instead of cancel, I don't want to set a data limit (please consult with a copy writer - not my best suit ;) ) 

Do you have a design mockup for xs?  My guess is that it's full screen with all the same functionality - but not clear unless we have it spec'd out.

**Comment by gkychow** [comment]
@DanGenuardi - Agree with what @twirlingsky commented on. 

A few other notes:
- Did we try a version (for desktop) which has a bit more padding or space around it? Because we're using 32px spacing around everything, it feels a bit too uniform.
- We should probably align with how "close" is displayed throughout the site for consistency.

**Comment by DanGenuardi** [comment]
@twirlingsky & @gkychow 

Thank you for your feedback and suggestions. I have taken some of those suggestions and implemented some of them into the updated modal designs.

1. I have changed the background colour to Shark Grey and lowered the opacity to 75%. I tried out the suggested 60% and I felt like the content behind the modal was still visible enough to distract you away from the modal itself. So I added the extra 15% to make it a little darker and make the modal box stand out more.

2. I added rounded corners to the modal box and made them 4px around.

3. I checked the TDS guidelines site and for the dismiss-able notification, we use an "x" icon to close the notification. So I recommend we keep the "x" as the close icon, and not add the word close next to it to avoid making the modal cluttered with clickable areas.
<img width="1004" alt="screen shot 2019-02-13 at 2 35 03 pm" src="https://user-images.githubusercontent.com/39997967/52744908-d7538980-2fab-11e9-98d3-6bf03c015862.png">

4. I changed the "x" icon colour from shark grey to purple.

5. I created a desktop view point, and made the padding 40px around for that and tablet view. I kept the mobile view padding set to 24pixels.

6. Width sizes for each breakpoint will be. 
Desktop - 770 pixels wide and the height varies on whatever content is in the modal.
Tablet - 480 pixels wide and the height varies on whatever content is in the modal.
Mobile - 288 pixels on a 320 pixel size screen. That will vary on the type of device resolution.

![screen shot 2019-02-13 at 4 28 21 pm](https://user-images.githubusercontent.com/39997967/52745249-94de7c80-2fac-11e9-8960-cf1aaf11681d.png)
![screen shot 2019-02-13 at 4 28 34 pm](https://user-images.githubusercontent.com/39997967/52745251-94de7c80-2fac-11e9-9a42-9b8881b51482.png)
![screen shot 2019-02-13 at 4 28 41 pm](https://user-images.githubusercontent.com/39997967/52745252-94de7c80-2fac-11e9-8bfb-2dccf3537381.png)

Modal Rules.

Here are some basic rules that I came up with. If anyone else has any to add please feel free to add.

1. Confirmation process within an action, "are you sure". We should include some important information within this modal. for example if a customer is deleting something from their account or making any changes within their account we want to make sure that they know what they're doing.

2. Data Manager page within Usage. We want to use a modal treatment for when someone wants to set an data limit to someone within their shared plan account.

3. Login treatment within the renewal flow.

DON'Ts
- don't use it for step process forms
- don't put to much information within the modal. like a sign up form with a lot of content to fill out.

**Comment by DanGenuardi** [comment]
@owestinTelus Could you please have a look at the most resent designs in this chain and let us know if the "x" icon would pass accessibility for closing the modal without the word close next to it?

**Comment by owestinTelus** [comment]
It's always beneficial to include the 'close' with the [x]. That said, the [x] is a recognizable design pattern and is passable. 

I'm more concerned about the need for you to use a modal at all. Most of the time an in page alert message or process step is a better solution. There is rarely a reason to interrupt the user flow. 

If the modal is the best design pattern, please research and document all of the interactions required to make the modal accessible. That way, you reduce the risk of poor dev implementation and provide the testers with the ability to validate that the interaction was implemented correctly. 

I've included a couple links, for consideration. 

**TPG - The current state of modal dialog accessibility**
https://developer.paciellogroup.com/blog/2018/06/the-current-state-of-modal-dialog-accessibility/

**W3 Aria-Practices example: Modal Dialog Example**
https://www.w3.org/TR/wai-aria-practices/examples/dialog-modal/dialog.html

**Comment by philroberto** [comment]
@owestinTelus We've made sure to include your recommendations around usage, considerations 
 and modal requirements ;-). I'd like those same considerations and usage to be echoed in the eventual documentation for this component.

@DanGenuardi thank you for doing your due diligence with this, it's great work. Please ensure that the usage and considerations up at the top of the page are referenced or included in any rules that are set for modals.

**Comment by varunj90** [comment]
hey @DanGenuardi based on the dpa meeting today
1. need to fix the x and close based on https://github.com/telus/tds-community/issues/175 (look at last comment)
2. mobile taking entire screen: https://www.w3.org/TR/wai-aria-practices/examples/dialog-modal/dialog.html
3. consider transition when motion and animation principles are available
4. shouldn't have access to back content of modal - high priority for accessibility

@philroberto can assist, thank you!

**Comment by apurvray** [comment]
Guidelines for modals and dialogue boxes .. in the document linked below

https://docs.google.com/document/d/1yLYnLHIaSVId7jlppsvhoS72PhzqyPd6FUVuCko1zAM/edit?usp=sharing

**Comment by donnavitan** [comment]
I wanted to include the following spinner design for consideration and as a reference to the treatment of the background overlay as approved by @gkychow: https://telus.invisionapp.com/share/4ARREY09XVM

Background overlay
* 100% height and width of viewport
* White (#FFFFFF)
* 94% Opacity

Example:
![image](https://user-images.githubusercontent.com/21316148/56899904-9aebe180-6a62-11e9-8ce9-d07a33acf4bd.png)

**Comment by ah-arch** [comment]
- A consideration would be to let the modal wrapper contain elements with flexible sizes rather than fixed responsive widths.

The overview app currently has a full-screen modal / overlay - with a X button that has hidden content reused from the link "_Continue in browser_":
![modal](https://user-images.githubusercontent.com/4450690/57389153-97093f00-7187-11e9-8fc2-469cc3413de1.png) ![wall](https://user-images.githubusercontent.com/4450690/57389073-72ad6280-7187-11e9-99fd-68a79ffa8b80.png)
- [Invision](https://telus.invisionapp.com/share/48QEGOZEJQH#/screens/345603814)
- Code: [Modal](https://github.com/telus/my-account-overview/blob/master/app/src/ui/components/Common/Modal/Modal.jsx) with nested [Wall](https://github.com/telus/my-account-overview/blob/master/app/src/ui/components/Wall/Wall.jsx#L37-L75) of `<Box vertical={5} horizontal={5} />`

The modal implements many of the [react-aria-modal](https://github.com/davidtheclark/react-aria-modal) accessibility features and there is [documentation](https://github.com/telus/business-account-ui/issues/22) for a 2017 version of it in the business-account-ui app:
![hub intro](https://user-images.githubusercontent.com/4450690/57390106-91acf400-7189-11e9-821c-c4ff4e9ada68.png)

**Comment by charliehay** [comment]
Would it be possible to change the "cancel" link to a button? The current design language states that links change the url, however in the case of a modal the url wouldn't change at all. I know we don't have a white button with gray text - but maybe this could be a solution as well?

**Comment by varunj90** [comment]
hey @apurvray  can you post the invision link here, thank you!

**Comment by apurvray** [comment]
Sketch symbols added to DSM Community (non-coded) along with guideline for using Content modal and dialog modal. 

The designs can be found here on invision for the devs working on them. 
https://telus.invisionapp.com/project/17778239

@enrico we can start working on the acceptance criteria now

**Comment by enrico** [comment]
FYI, wrong Enrico, cc'ing @theetrain

**Comment by varunj90** [comment]
this is now being story written by @theetrain

**Comment by marcod1419** [comment]
Hi @dustin-ruetz! I've looked it over, and I agree that this could be a good starting point! Once we begin development, we can give it another look and hopefully use it to get us a bit ahead of the game.

For those that will be working with it, there are a few small things that should be addressed that I've found during a preliminary lookover to match up with our specific ACs:

- The close button is focused first, but that also applies when there’s a heading. The heading should be focused first if available.
- The header is a span, not an h3
- The contents tend to not adjust when the page size changes (see text modal while open)
- The buttons need alignment to match our design
- While accessible via voice over, it will read all the contents of the modal out at once. This can be improved.

**Comment by varunj90** [comment]
Development is almost completed by @nicmak and will move into QA phase shortly after

**Comment by varunj90** [comment]
@nicmak will be working on converting this to react hooks this week

**Comment by affansajid** [comment]
@nicmak I have some notes from QAing the model based on the Acceptance criteria, you can find them here: https://docs.google.com/document/d/1_uAfGEn58GyHQxcaCesSSlhZFlM_s-Duk1iuTYH5vwc/edit?usp=sharing

**Comment by twirlingsky** [comment]
Home Solutions is working on putting PDP information into a modal. Here is a link to the screens:
XL: https://telus.invisionapp.com/d/main#/console/19265569/405479480/preview
XS: https://telus.invisionapp.com/d/main#/console/19327711/405477181/preview

(Just wanting to have these linked to the ticket for now)

**Comment by twirlingsky** [comment]
Did we change to Athens Grey for the overlay colour? I thought we decided on Shark Grey. Athens will be hard on low contrast monitors

**Comment by donnavitan** [comment]
Hello @twirlingsky , Shark Grey was used in previous designs but the design ACs and Invision links should be using the latest specs: https://telus.invisionapp.com/d/main#/projects/prototypes/17778239

We had discussions with @gkychow and @apurvray to use the Athens Gray at 94% opacity to align with the Brand aesthetic. It's not intended to make the modal card container stand out, but rather make the whole view very washed out to let the content of the modal be the focus.

Let me know if that helps clarify things.

**Comment by theetrain** [comment]
`@tds/community-modal@1.0.0` has been released. 
It is also available in TDS_Community DSM v1.0.13

Thank you for your contributions!
